File: Acupay/Core Forms/Salary/Salary.vb
Imports System.ComponentModel.DataAnnotations.Schema
Imports System.ComponentModel.DataAnnotations

Namespace Global.PayrollSys

    <Table("employeesalary")>
    Public Class Salary

        <Key>
        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property RowID As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Identity)>
        Public Property Created As Date

        Public Property CreatedBy As Integer?

        <DatabaseGenerated(DatabaseGeneratedOption.Computed)>
        Public Property LastUpd As Date?

        Public Property LastUpdBy As Integer?

        Public Property EmployeeID As Integer?

        Public Property OrganizationID As Integer?

        Public Property FilingStatusID As Integer?

        Public Property PositionID As Integer?

        Public Property PaySocialSecurityID As Integer?

        Public Property PayPhilHealthID As Integer?

        Public Property PhilHealthDeduction As Decimal

        Public Property HDMFAmount As Decimal

        Public Property BasicPay As Decimal

        <Column("Salary")>
        Public Property BasicSalary As Decimal

        <Column("UndeclaredSalary")>
        Public Property AllowanceSalary As Decimal

        <Column("TrueSalary")>
        Public Property TotalSalary As Decimal

        <Column("BasicDailyPay")>
        Public Property DailyRate As Decimal

        <Column("BasicHourlyPay")>
        Public Property HourlyRate As Decimal

        Public Property NoOfDependents As Integer

        Public Property MaritalStatus As String

        <Column("EffectiveDateFrom")>
        Public Property EffectiveFrom As Date

        <Column("EffectiveDateTo")>
        Public Property EffectiveTo As Date?

        Public Property OverrideDiscardSSSContrib As Boolean

        Public Property OverrideDiscardPhilHealthContrib As Boolean

        Public ReadOnly Property SSSDeduction As Decimal
            Get
                Return If(SocialSecurityBracket?.EmployeeContributionAmount, 0)
            End Get
        End Property

        Public ReadOnly Property IsIndefinite As Boolean
            Get
                Return Not EffectiveTo.HasValue
            End Get
        End Property

        <ForeignKey("PaySocialSecurityID")>
        Public Overridable Property SocialSecurityBracket As SocialSecurityBracket

    End Class

End Namespace

File: AccuPay.Core/Entities/Salary.cs
using System;
using System.ComponentModel.DataAnnotations.Schema;
using System.Globalization;

namespace AccuPay.Core.Entities
{
    [Table("employeesalary")]
    public partial class Salary : EmployeeDataEntity
    {
        public int? PositionID { get; set; }

        public decimal PhilHealthDeduction { get; set; }

        public decimal HDMFAmount { get; set; }

        [Column("Salary")]
        public decimal BasicSalary { get; set; }

        [Column("UndeclaredSalary")]
        public decimal AllowanceSalary { get; set; }

        // TODO: make this readonly. Domain methods should only be the one to set this.
        [Column("TrueSalary")]
        public decimal TotalSalary { get; internal set; }

        public string MaritalStatus { get; set; }

        [Column("EffectiveDateFrom")]
        public DateTime EffectiveFrom { get; set; }

        [ForeignKey("EmployeeID")]
        public virtual Employee Employee { get; set; }

        public bool DoPaySSSContribution { get; set; }

        public bool AutoComputePhilHealthContribution { get; set; }

        public bool AutoComputeHDMFContribution { get; set; }

        public bool IsMinimumWage { get; set; }
    }

    public partial class Salary
    {
        public Salary()
        {
        }

        public Salary(int? organizationId,
            int? employeeId,
            int? positionId,
            DateTime effectiveFrom,
            decimal basicSalary,
            decimal allowanceSalary,
            bool doPaySSSContribution,
            bool autoComputeHDMFContribution,
            bool autoComputePhilHealthContribution,
            decimal hdmfAmount,
            decimal philHealthDeduction)
        {
            OrganizationID = organizationId;
            EmployeeID = employeeId;
            PositionID = positionId;
            EffectiveFrom = effectiveFrom;
            BasicSalary = basicSalary;
            AllowanceSalary = allowanceSalary;
            DoPaySSSContribution = doPaySSSContribution;
            AutoComputeHDMFContribution = autoComputeHDMFContribution;
            AutoComputePhilHealthContribution = autoComputePhilHealthContribution;
            HDMFAmount = hdmfAmount;
            PhilHealthDeduction = philHealthDeduction;
        }

        public static Salary NewSalary(int? organizationId,
            int? employeeId,
            int? positionId,
            DateTime effectiveFrom,
            decimal basicSalary,
            decimal allowanceSalary,
            bool doPaySSSContribution,
            bool autoComputeHDMFContribution,
            bool autoComputePhilHealthContribution,
            decimal hdmfAmount,
            decimal philHealthDeduction) => new Salary(organizationId: organizationId,
                employeeId: employeeId,
                positionId: positionId,
                effectiveFrom: effectiveFrom,
                basicSalary: basicSalary,
                allowanceSalary: allowanceSalary,
                doPaySSSContribution: doPaySSSContribution,
                autoComputeHDMFContribution: autoComputeHDMFContribution,
                autoComputePhilHealthContribution: autoComputePhilHealthContribution,
                hdmfAmount: hdmfAmount,
                philHealthDeduction: philHealthDeduction);

        /// <summary>
        /// Updates TotalSalary. Call this everytime BasicSalary or AllowanceSalary has changed.
        /// </summary>
        public void UpdateTotalSalary() => TotalSalary = BasicSalary + AllowanceSalary;

        public string AutoComputeSSSDisplay => DoPaySSSContribution ? "AUTO" : "NO";

        public string AutoComputePhilHealthDisplay => AutoComputePhilHealthContribution ? "AUTO" : PhilHealthDeduction.ToString("N", CultureInfo.CurrentCulture);

        public string AutoComputeHDMFDisplay => AutoComputeHDMFContribution ? "AUTO" : HDMFAmount.ToString("N", CultureInfo.CurrentCulture);
    }
}
